include/SampleCache.h

```diff
@@ -98,23 +98,16 @@ class SampleCache
 	static auto get(const T& entry, std::unordered_map<T, std::weak_ptr<SampleBuffer>, Hash>& map, Args... args)
 	{
 		const auto it = map.find(entry);
+		const auto item = it == map.end() ? nullptr : it->second.lock();
```

Comment:
This last commit didn't fix either of the problems I mentioned. `entry` still has an extra copy and the map iterator is not reused so a lookup occurs twice (once in `find` and again in `emplace`).

I was thinking something like this:
```cpp
template <typename T, typename... Args>
static auto get(T entry, std::unordered_map<T, std::weak_ptr<SampleBuffer>, Hash>& map, Args&&... args)
{
	const auto it = map.find(entry);

	if (it == map.end())
	{
		const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
		map.emplace(std::move(entry), buffer);
		return buffer;
	}

	const auto entryLock = it->second.lock();
	if (!entryLock)
	{
		const auto buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
		it->second = buffer;
		return buffer;
	}

	return entryLock;
}
```

Calling code (just `SampleCache::fetch` I think) should move `entry` when calling.

I also made the `args` parameter an rvalue reference because that is needed for perfect forwarding.

Reply:
Oops, my bad, fixing now.

Reply:
Wait, `emplace` does a lookup?

Reply:
I actually think we can use `insert_or_assign` here.

Reply:
I don't think `insert_or_assign` would be useful in this situation. The "insert" part is good, but you don't want to unconditionally assign to the map since you need to call the `lock` method on the weak pointer and return the result without reassigning it if it is non-null.

Reply:
The issues I mentioned still aren't solved. `insert_or_assign` still needs to look up the key, and you shouldn't call `std::move` on a const reference.

The code I shared in my previous comment fixes both. However, while it did fix the unnecessary lookup in the `!entryLock` branch, there was still a 2nd lookup in the `it == map.end()` branch.

But now I think I've fixed that as well, while also simplifying the code:

```cpp
// NOTE: `entry` is a value not a reference, and calling code can use `std::move`
template <typename T, typename... Args>
static auto get(T entry, std::unordered_map<T, std::weak_ptr<SampleBuffer>, Hash>& map, Args&&... args)
{
	auto& value = map[std::move(entry)]; // NOTE: This is the only lookup; moves from `entry` and default constructs value if key does not exist

	auto buffer = value.lock(); // NOTE: null if key didn't exist or previous buffer was freed
	if (!buffer)
	{
		buffer = std::make_shared<SampleBuffer>(std::forward<Args>(args)...);
		value = buffer;
	}

	return buffer;
}
```

The `NOTE` comments are added for explanation.